lib/DxilPIXPasses/DxilPIXDXRInvocationsLog.cpp

```diff
@@ -0,0 +1,234 @@
+//                                                                           //
+// DxilPIXDXRInvocationsLog.cpp                                              //
+// Copyright (C) Microsoft Corporation. All rights reserved.                 //
+// This file is distributed under the University of Illinois Open Source     //
+// License. See LICENSE.TXT for details.                                     //
+//                                                                           //
+///////////////////////////////////////////////////////////////////////////////
+
+#include "dxc/DXIL/DxilFunctionProps.h"
+#include "dxc/DXIL/DxilOperations.h"
+#include "dxc/DXIL/DxilUtil.h"
+
+#include "dxc/DXIL/DxilInstructions.h"
+#include "dxc/DXIL/DxilModule.h"
+#include "dxc/DxilPIXPasses/DxilPIXPasses.h"
+
+#include "llvm/IR/InstIterator.h"
+#include "llvm/IR/PassManager.h"
+#include "llvm/Support/FormattedStream.h"
+#include "llvm/Transforms/Utils/Local.h"
+
+#include "PixPassHelpers.h"
+
+using namespace llvm;
+using namespace hlsl;
+using namespace PIXPassHelpers;
+
+class DxilPIXDXRInvocationsLog : public ModulePass {
+    uint64_t m_MaxNumEntriesInLog = 10;
+public:
+  static char ID;
+  DxilPIXDXRInvocationsLog() : ModulePass(ID) {}
+  StringRef getPassName() const override {
+    return "DXIL Logs all non-RayGen DXR 1.0 invocations into a UAV";
+  }
+
+  void applyOptions(PassOptions O) override;
+  bool runOnModule(Module &M) override;
+};
+
+static DXIL::ShaderKind GetShaderKind(DxilModule const &DM, llvm::Function const *entryFunction)
+{
+    DXIL::ShaderKind ShaderKind = DXIL::ShaderKind::Invalid;
+    if (!DM.HasDxilFunctionProps(entryFunction)) {
+        auto ShaderModel = DM.GetShaderModel();
+        ShaderKind = ShaderModel->GetKind();
+    }
+    else {
+        auto const& Props = DM.GetDxilFunctionProps(entryFunction);
+        ShaderKind = Props.shaderKind;
+    }
+
+    return ShaderKind;
+}
+
+void DxilPIXDXRInvocationsLog::applyOptions(PassOptions O) {
+    GetPassOptionUInt64(O, "maxNumEntriesInLog", &m_MaxNumEntriesInLog, 10); // Use a silly default value. PIX should set a better value here.
+}
+
+bool DxilPIXDXRInvocationsLog::runOnModule(Module &M) {
+
+  DxilModule &DM = M.GetOrCreateDxilModule();
+  LLVMContext &Ctx = M.getContext();
+  OP *HlslOP = DM.GetOP();
+
+  for (auto entryFunction : DM.GetExportedFunctions()) {
+    
+    DXIL::ShaderKind ShaderKind = GetShaderKind(DM, entryFunction);
+    
+    switch (ShaderKind) {
+    case DXIL::ShaderKind::Intersection:
+    case DXIL::ShaderKind::AnyHit:
+    case DXIL::ShaderKind::ClosestHit:
+    case DXIL::ShaderKind::Miss:
+      break;
+    
+    default:
+      continue;
+    }
+
+    IRBuilder<> Builder(dxilutil::FirstNonAllocaInsertionPt(entryFunction));
+
+    // Add the UAVs that we're going to write to
+    CallInst* HandleForCountUAV = PIXPassHelpers::CreateUAV(DM, Builder, 0, "PIX_CountUAV_Handle");
```

Comment:
Should the 0 and the 1 on the next line be put in a constant somewhere? I see it repeated in a few places.